File: backend/src/test/java/com/github/aadvorak/artilleryonline/ContactUtilsTest.java
package com.github.aadvorak.artilleryonline;

import com.github.aadvorak.artilleryonline.battle.common.BodyPosition;
import com.github.aadvorak.artilleryonline.battle.common.Position;
import com.github.aadvorak.artilleryonline.battle.common.lines.Circle;
import com.github.aadvorak.artilleryonline.battle.common.lines.HalfCircle;
import com.github.aadvorak.artilleryonline.battle.common.lines.Trapeze;
import com.github.aadvorak.artilleryonline.battle.common.shapes.TrapezeShape;
import com.github.aadvorak.artilleryonline.battle.utils.ContactUtils;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class ContactUtilsTest {

    private static final double SMALL_DELTA = 0.00001;

    private static final double SQRT_05 = Math.sqrt(0.5);

    private static final TrapezeShape RECT_TRAPEZE_SHAPE = new TrapezeShape()
            .setBottomRadius(1.0)
            .setTopRadius(1.0)
            .setHeight(1.0);

    private static final TrapezeShape SMALLER_TOP_TRAPEZE_SHAPE = new TrapezeShape()
            .setBottomRadius(1.0)
            .setTopRadius(0.4)
            .setHeight(1.0);

    @Test
    public void trapezesBottomWithBottomLeft() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setX(-0.5)
                        .setY(-0.5)
                        .setAngle(-3 * Math.PI / 4),
                RECT_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(1.0, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesRightWithTopLeft() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setX(2.2)
                        .setAngle(Math.PI / 8),
                RECT_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(-1.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(0.0, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesTopWithBottomRight() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setY(1.2)
                        .setAngle(-Math.PI / 8),
                RECT_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-1.0, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesLeftWithTopRight() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setX(-2.2)
                        .setAngle(-Math.PI / 8),
                RECT_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(1.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(0.0, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesBottomWithTop() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setY(-0.9),
                SMALLER_TOP_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(1.0, contact.normal().getY(), SMALL_DELTA),
                () -> assertEquals(0.1, contact.depth(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesRightWithTop() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setX(1.9)
                        .setY(0.5)
                        .setAngle(Math.PI / 2),
                SMALLER_TOP_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(-1.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(0.0, contact.normal().getY(), SMALL_DELTA),
                () -> assertEquals(0.1, contact.depth(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesTopRightWithBottomLeft() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setX(-1.9)
                        .setY(-0.9),
                RECT_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(SQRT_05, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(SQRT_05, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesTopLeftWithBottomRight() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setX(1.9)
                        .setY(-0.9),
                RECT_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(-SQRT_05, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(SQRT_05, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesBottomLeftWithTopRight() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setX(1.9)
                        .setY(0.9),
                RECT_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(-SQRT_05, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-SQRT_05, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezesBottomRightWithTopLeft() {
        var trapeze = new Trapeze(
                new BodyPosition()
                        .setX(-1.9)
                        .setY(0.9),
                RECT_TRAPEZE_SHAPE
        );
        var otherTrapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(SQRT_05, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-SQRT_05, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeBottomRightWithCircle() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var circle = new Circle(new Position().setX(1.5).setY(-0.5), 1.0);
        var contact = ContactUtils.getTrapezeCircleContact(trapeze, circle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(SQRT_05, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-SQRT_05, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeBottomWithCircle() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var circle = new Circle(new Position().setY(-0.5), 1.0);
        var contact = ContactUtils.getTrapezeCircleContact(trapeze, circle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-1.0, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeBottomWithCircleCenterInTrapeze() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var circle = new Circle(new Position().setY(0.1), 0.2);
        var contact = ContactUtils.getTrapezeCircleContact(trapeze, circle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-1.0, contact.normal().getY(), SMALL_DELTA),
                () -> assertEquals(0.3, contact.depth(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeBottomWithCircleFullInTrapeze() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var circle = new Circle(new Position().setY(0.11), 0.1);
        var contact = ContactUtils.getTrapezeCircleContact(trapeze, circle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-1.0, contact.normal().getY(), SMALL_DELTA),
                () -> assertEquals(0.21, contact.depth(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeWithCircleNoContact() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var circle = new Circle(new Position().setY(-0.11), 0.1);
        var contact = ContactUtils.getTrapezeCircleContact(trapeze, circle);
        assertNull(contact);
    }

    @Test
    public void trapezeWithHalfCircleNoContact() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var halfCircle = HalfCircle.of(new BodyPosition().setY(RECT_TRAPEZE_SHAPE.getHeight()), 1.0);
        var contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle);
        assertNull(contact);
    }

    @Test
    public void trapezeTopWithHalfCircleBottom() {
        var depth = 0.1;
        var trapeze = new Trapeze(
                new BodyPosition().setY(depth),
                SMALLER_TOP_TRAPEZE_SHAPE
        );
        var halfCircle = HalfCircle.of(new BodyPosition().setY(RECT_TRAPEZE_SHAPE.getHeight()), 1.0);
        var contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(1.0, contact.normal().getY(), SMALL_DELTA),
                () -> assertEquals(depth, contact.depth(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeTopWithHalfCircleTop() {
        var depth = 0.1;
        var trapeze = new Trapeze(
                new BodyPosition().setY(depth),
                SMALLER_TOP_TRAPEZE_SHAPE
        );
        var halfCircle = HalfCircle.of(new BodyPosition()
                .setY(RECT_TRAPEZE_SHAPE.getHeight() + 1.0)
                .setAngle(Math.PI), 1.0);
        var contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(1.0, contact.normal().getY(), SMALL_DELTA),
                () -> assertEquals(depth, contact.depth(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeBottomWithHalfCircleTop() {
        var depth = 0.1;
        var trapeze = new Trapeze(
                new BodyPosition().setY(-depth),
                SMALLER_TOP_TRAPEZE_SHAPE
        );
        var halfCircle = HalfCircle.of(new BodyPosition().setY(-1.0), 1.0);
        var contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(0.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-1.0, contact.normal().getY(), SMALL_DELTA),
                () -> assertEquals(depth, contact.depth(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeBottomRightWithHalfCircleTop() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var halfCircle = HalfCircle.of(new BodyPosition()
                .setX(1.5)
                .setY(-0.5)
                .setAngle(Math.PI / 4), 1.0);
        var contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(SQRT_05, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(-SQRT_05, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeTopRightWithHalfCircleBottom() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var halfCircle = HalfCircle.of(new BodyPosition()
                .setX(0.9)
                .setY(0.9)
                .setAngle(-Math.PI / 4), 1.0);
        var contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(SQRT_05, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(SQRT_05, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeRightWithHalfCircleLeftBottom() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var halfCircle = HalfCircle.of(new BodyPosition()
                .setX(1.9)
                .setY(0.5), 1.0);
        var contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(1.0, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(0.0, contact.normal().getY(), SMALL_DELTA)
        );
    }

    @Test
    public void trapezeTopRightWithHalfCircleLeftBottom() {
        var trapeze = new Trapeze(
                new BodyPosition(),
                RECT_TRAPEZE_SHAPE
        );
        var halfCircle = HalfCircle.of(new BodyPosition()
                .setX(1.9)
                .setY(0.9), 1.0);
        var contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle);
        assertNotNull(contact);
        assertAll(
                () -> assertEquals(SQRT_05, contact.normal().getX(), SMALL_DELTA),
                () -> assertEquals(SQRT_05, contact.normal().getY(), SMALL_DELTA)
        );
    }
}

File: frontend/test/contact-utils.test.ts
import {ContactUtils} from "~/playground/utils/contact-utils"
import {Trapeze, Circle, HalfCircle} from "~/playground/data/geometry"
import {test, expect} from "@jest/globals"

const SQRT_05 = Math.sqrt(0.5)

const RECT_TRAPEZE_SHAPE = {
  name: "Trapeze",
  bottomRadius: 1.0,
  topRadius: 1.0,
  height: 1.0
}

const SMALLER_TOP_TRAPEZE_SHAPE = {
  name: "Trapeze",
  bottomRadius: 1.0,
  topRadius: 0.4,
  height: 1.0
}

test('trapezesBottomWithBottomLeft', () => {
  const trapeze = new Trapeze(
    {
      x: -0.5,
      y: -0.5,
      angle: -3 * Math.PI / 4
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(1.0, 5)
  }
})

test('trapezesRightWithTopLeft', () => {
  const trapeze = new Trapeze(
    {
      x: 2.2,
      y: 0,
      angle: Math.PI / 8
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(-1.0, 5)
    expect(contact.normal.y).toBeCloseTo(0.0, 5)
  }
})

test('trapezesTopWithBottomRight', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 1.2,
      angle: -Math.PI / 8
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(-1.0, 5)
  }
})

test('trapezesLeftWithTopRight', () => {
  const trapeze = new Trapeze(
    {
      x: -2.2,
      y: 0,
      angle: -Math.PI / 8
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(1.0, 5)
    expect(contact.normal.y).toBeCloseTo(0.0, 5)
  }
})

test('trapezesBottomWithTop', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: -0.9,
      angle: 0
    },
    SMALLER_TOP_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(1.0, 5)
    expect(contact.depth).toBeCloseTo(0.1, 5)
  }
})

test('trapezesRightWithTop', () => {
  const trapeze = new Trapeze(
    {
      x: 1.9,
      y: 0.5,
      angle: Math.PI / 2
    },
    SMALLER_TOP_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(-1.0, 5)
    expect(contact.normal.y).toBeCloseTo(0.0, 5)
    expect(contact.depth).toBeCloseTo(0.1, 5)
  }
})

test('trapezesTopRightWithBottomLeft', () => {
  const trapeze = new Trapeze(
    {
      x: -1.9,
      y: -0.9,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(SQRT_05, 5)
    expect(contact.normal.y).toBeCloseTo(SQRT_05, 5)
  }
})

test('trapezesTopLeftWithBottomRight', () => {
  const trapeze = new Trapeze(
    {
      x: 1.9,
      y: -0.9,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(-SQRT_05, 5)
    expect(contact.normal.y).toBeCloseTo(SQRT_05, 5)
  }
})

test('trapezesBottomLeftWithTopRight', () => {
  const trapeze = new Trapeze(
    {
      x: 1.9,
      y: 0.9,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(-SQRT_05, 5)
    expect(contact.normal.y).toBeCloseTo(-SQRT_05, 5)
  }
})

test('trapezesBottomWithTop1', () => {
  const trapeze = new Trapeze(
    {
      x: 1.7,
      y: 0.9,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(-1.0, 5)
    expect(contact.depth).toBeCloseTo(0.1, 5)
  }
})

test('trapezesLeftWithRight', () => {
  const trapeze = new Trapeze(
    {
      x: 1.9,
      y: 0.7,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(-1.0, 5)
    expect(contact.normal.y).toBeCloseTo(0.0, 5)
    expect(contact.depth).toBeCloseTo(0.1, 5)
  }
})

test('trapezesBottomRightWithTopLeft', () => {
  const trapeze = new Trapeze(
    {
      x: -1.9,
      y: 0.9,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const otherTrapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const contact = ContactUtils.getTrapezesContact(trapeze, otherTrapeze)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(SQRT_05, 5)
    expect(contact.normal.y).toBeCloseTo(-SQRT_05, 5)
  }
})

test('trapezeBottomRightWithCircle', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const circle = new Circle({x: 1.5, y: -0.5}, 1.0)
  const contact = ContactUtils.getTrapezeCircleContact(trapeze, circle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(SQRT_05, 5)
    expect(contact.normal.y).toBeCloseTo(-SQRT_05, 5)
  }
})

test('trapezeBottomWithCircle', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const circle = new Circle({x: 0, y: -0.5}, 1.0)
  const contact = ContactUtils.getTrapezeCircleContact(trapeze, circle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(-1.0, 5)
  }
})

test('trapezeBottomWithCircleCenterInTrapeze', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const circle = new Circle({x: 0, y: 0.1}, 0.2)
  const contact = ContactUtils.getTrapezeCircleContact(trapeze, circle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(-1.0, 5)
    expect(contact.depth).toBeCloseTo(0.3, 5)
  }
})

test('trapezeBottomWithCircleFullInTrapeze', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const circle = new Circle({x: 0, y: 0.11}, 0.1)
  const contact = ContactUtils.getTrapezeCircleContact(trapeze, circle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(-1.0, 5)
    expect(contact.depth).toBeCloseTo(0.21, 5)
  }
})

test('trapezeWithCircleNoContact', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const circle = new Circle({x: 0, y: -0.11}, 0.1)
  const contact = ContactUtils.getTrapezeCircleContact(trapeze, circle)
  expect(contact).toBeNull()
})

test('trapezeWithHalfCircleNoContact', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const halfCircle = HalfCircle.of({x: 0, y: RECT_TRAPEZE_SHAPE.height, angle: 0}, 1.0)
  const contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle)
  expect(contact).toBeNull()
})

test('trapezeTopWithHalfCircleBottom', () => {
  const depth = 0.1
  const trapeze = new Trapeze(
    {
      x: 0,
      y: depth,
      angle: 0
    },
    SMALLER_TOP_TRAPEZE_SHAPE
  )
  const halfCircle = HalfCircle.of({x: 0, y: RECT_TRAPEZE_SHAPE.height, angle: 0}, 1.0)
  const contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(1.0, 5)
    expect(contact.depth).toBeCloseTo(depth, 5)
  }
})

test('trapezeTopWithHalfCircleTop', () => {
  const depth = 0.1
  const trapeze = new Trapeze(
    {
      x: 0,
      y: depth,
      angle: 0
    },
    SMALLER_TOP_TRAPEZE_SHAPE
  )
  const halfCircle = HalfCircle.of({x: 0, y: RECT_TRAPEZE_SHAPE.height + 1.0, angle: Math.PI}, 1.0)
  const contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(1.0, 5)
    expect(contact.depth).toBeCloseTo(depth, 5)
  }
})

test('trapezeBottomWithHalfCircleTop', () => {
  const depth = 0.1
  const trapeze = new Trapeze(
    {
      x: 0,
      y: -depth,
      angle: 0
    },
    SMALLER_TOP_TRAPEZE_SHAPE
  )
  const halfCircle = HalfCircle.of({x: 0, y: -1.0, angle: 0}, 1.0)
  const contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(0.0, 5)
    expect(contact.normal.y).toBeCloseTo(-1.0, 5)
    expect(contact.depth).toBeCloseTo(depth, 5)
  }
})

test('trapezeBottomRightWithHalfCircleTop', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const halfCircle = HalfCircle.of({x: 1.5, y: -0.5, angle: Math.PI / 4}, 1.0)
  const contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(SQRT_05, 5)
    expect(contact.normal.y).toBeCloseTo(-SQRT_05, 5)
  }
})

test('trapezeTopRightWithHalfCircleBottom', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const halfCircle = HalfCircle.of({x: 0.9, y: 0.9, angle: -Math.PI / 4}, 1.0)
  const contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(SQRT_05, 5)
    expect(contact.normal.y).toBeCloseTo(SQRT_05, 5)
  }
})

test('trapezeRightWithHalfCircleLeftBottom', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const halfCircle = HalfCircle.of({x: 1.9, y: 0.5, angle: 0}, 1.0)
  const contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(1.0, 5)
    expect(contact.normal.y).toBeCloseTo(0.0, 5)
  }
})

test('trapezeTopRightWithHalfCircleLeftBottom', () => {
  const trapeze = new Trapeze(
    {
      x: 0,
      y: 0,
      angle: 0
    },
    RECT_TRAPEZE_SHAPE
  )
  const halfCircle = HalfCircle.of({x: 1.9, y: 0.9, angle: 0}, 1.0)
  const contact = ContactUtils.getTrapezeHalfCircleContact(trapeze, halfCircle)
  expect(contact).not.toBeNull()
  if (contact) {
    expect(contact.normal.x).toBeCloseTo(SQRT_05, 5)
    expect(contact.normal.y).toBeCloseTo(SQRT_05, 5)
  }
})